src/ui-scroll.directive.ts

```diff
@@ -20,13 +21,19 @@ export class UiScrollDirective implements OnInit {
     this.datasource = datasource;
   }
 
+  @Input() set uiScrollTable(value: any) {
+    this.isTable = !!value;
+  }
+
   ngOnInit() {
     const compFactory = this.resolver.resolveComponentFactory(UiScrollComponent);
-    const componentRef = this.viewContainer.createComponent(
-      compFactory, undefined, this.viewContainer.injector
-    );
+    const componentRef = compFactory.create(this.viewContainer.injector);
     componentRef.instance.datasource = this.datasource;
     componentRef.instance.template = this.templateRef;
+    componentRef.instance.isTable = this.isTable;
     componentRef.instance.version = version;
+    componentRef.instance.parentElement = this.templateRef.elementRef.nativeElement.parentElement;
+    this.viewContainer.createEmbeddedView(componentRef.instance.uiScrollTemplateRef);
+    setTimeout(() => componentRef.changeDetectorRef.detectChanges()); // 😢
```

Comment:
@markdBC I admire your deep insight of the Angular internal things and very grateful for your explanation regarding ngOnInit. As you can see I had to wrap changes detection into another `setTimeout`, the problem is that the `onInit` hook is being triggered before the initial content of the component is rendered. Do you know, what can be done here? I tried replace `OnInit` with `AfterViewInit` (the latest init hook), but it also runs too early.

The initialization, I need to run on the components first render, includes DOM parsing, in particular, the Scroller tries to find elements needed for virtualization. And immediate initialization (before DOM gets necessary elements during first render) just fails.

Also, calling `componentRef.changeDetectorRef.detectChanges()` seems equal to calling `componentRef.instance.ngOnInit()`, but I agree this is better than calling `this.ngOnInit()` on the component's constructor.

That's what we have now... And it would be great to dig a little dipper and find a way to catch the time of the first render properly. The last thing I want to add is that the demo starts working, not perfect, but something can be seen running localhost:4200/#/test